File: judgesystem/judge/src/ftp.cpp
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
/*                                 _ooOoo_                                   */
/*                                o8888888o                                  */
/*                                88" . "88                                  */
/*                                (| -_- |)                                  */
/*                                O\  =  /O                                  */
/*                             ____/`---'\____                               */
/*                           .'  \\|     |//  `.                             */
/*                          /  \\|||  :  |||//  \                            */
/*                         /  _||||| -:- |||||-  \                           */
/*                         |   | \\\  -  /// |   |                           */
/*                         | \_|  ''\---/''  |   |                           */
/*                         \  .-\__  `-`  ___/-. /                           */
/*                       ___`. .'  /--.--\  `. . __                          */
/*                    ."" '< `.___\_<|>_/___.'  >'"".                        */
/*                   | | :  `- \`.;`\ _ /`;.`/ - ` : | |                     */
/*                   \  \ `-.   \_ __\ /__ _/   .-` /  /                     */
/*              ======`-.____`-.___\_____/___.-`____.-'======                */
/*                                 `=---='                                   */
/*              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                */
/*                       佛祖保佑       永無bug                              */
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
#include "ftp.h"

FTP::FTP(){
}
FTP::FTP(std::string _server, std::string _user, std::string _password, std::string _ftp_log){
    m_server = _server;
    m_user = _user;
    m_password = _password;
    m_log = LOG(_ftp_log);
    std::cout << m_server << ' ' << m_user << ' ' << m_password << std::endl;
}
bool FTP::get(std::string from, std::string to){
    std::cout << "Get: " << from << " to " << to << std::endl;
}

bool FTP::put(std::string from, std::string to){
    std::cout << "Put: " << from << " to " << to << std::endl;
}

File: judgesystem/judge/ftp.py
import paramiko, sys
from scp import SCPClient
import os
import time
import datetime
import sys

class FTP:
    def __init__(self, server, port, user, password):
        self.server = server
        self.port = port
        self.user = user
        self.password = password
        self.createSSHClient()

    def createSSHClient(self):
        self.client = paramiko.SSHClient()
        self.client.load_system_host_keys()
        self.client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
        self.client.connect(self.server, self.port, self.user, self.password)
    
    def get(self, _from, _to):
        while True:
            try:
                scp = SCPClient(self.client.get_transport())
                try: os.makedirs(os.path.split(_to)[0])
                except: pass
                scp.get(_from, _to, recursive=True)
                scp.close()
                return
            except:
                pass

    def put(self, _from, _to):
        try: ssh.exec_command('mkdir -p %s' % os.path.split(_to)[0])
        except: pass
        scp = SCPClient(self.client.get_transport())
        scp.put(_from, _to)
        scp.close()

    def delete(self, _target):
        try: self.client.exec_command('rm -rf %s' % _target)
        except: pass

if __name__ == "__main__":
    ftp = FTP("localhost", 22, "nctuojftp", "nctuojftp")
    print("sleep")
    time.sleep(100)
    ftp.put("./meta", "./")



    pass
